File: service/app/src/main/java/io/miragon/blueprint/adapter/inbound/rest/GlobalExceptionConfiguration.java
package io.miragon.blueprint.adapter.inbound.rest;

import java.net.URI;

import org.jspecify.annotations.Nullable;
import org.operaton.bpm.engine.MismatchingMessageCorrelationException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

/**
 * Cross-cutting web error handling for the REST adapter: turns domain and input errors into RFC 9457
 * `application/problem+json` responses (Spring serialises {@link ProblemDetail} as such automatically).
 *
 * This is web configuration, hence the `Configuration` suffix whitelisted for `adapter.inbound.rest`
 * and its home here rather than in a new top-level `config` package (which the architecture tests
 * would reject).
 */
@RestControllerAdvice
public class GlobalExceptionConfiguration {

    private static final Logger log = LoggerFactory.getLogger(GlobalExceptionConfiguration.class);

    /** Invalid input — a bad UUID, a blank value object, an unknown status filter. */
    @ExceptionHandler(IllegalArgumentException.class)
    public ProblemDetail handleIllegalArgument(IllegalArgumentException exception) {
        log.debug("Bad request: {}", exception.getMessage());
        return problem(HttpStatus.BAD_REQUEST, "Invalid request", exception.getMessage());
    }

    /** Unknown aggregate — the services signal a missing application with {@link IllegalStateException}. */
    @ExceptionHandler(IllegalStateException.class)
    public ProblemDetail handleIllegalState(IllegalStateException exception) {
        log.debug("Unprocessable: {}", exception.getMessage());
        return problem(HttpStatus.NOT_FOUND, "Resource not found", exception.getMessage());
    }

    /** Action not available — the process token is no longer at the expected wait state. */
    @ExceptionHandler(MismatchingMessageCorrelationException.class)
    public ProblemDetail handleMismatchingCorrelation(MismatchingMessageCorrelationException exception) {
        log.debug("Mismatching correlation: {}", exception.getMessage());
        return problem(HttpStatus.CONFLICT, "Action not available in the current state", exception.getMessage());
    }

    private ProblemDetail problem(HttpStatus status, String title, @Nullable String detail) {
        ProblemDetail problemDetail = ProblemDetail.forStatus(status);
        problemDetail.setTitle(title);
        problemDetail.setDetail(detail);
        problemDetail.setType(URI.create("https://miravelo.example/problems/" + status.value()));
        return problemDetail;
    }
}

File: service/app/src/main/kotlin/io/miragon/blueprint/adapter/inbound/rest/GlobalExceptionConfiguration.kt
package io.miragon.blueprint.adapter.inbound.rest

import mu.KotlinLogging
import org.operaton.bpm.engine.MismatchingMessageCorrelationException
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.net.URI

/**
 * Cross-cutting web error handling for the REST adapter: turns domain and input errors into RFC 9457
 * `application/problem+json` responses (Spring serialises [ProblemDetail] as such automatically).
 *
 * This is web configuration, hence the `Configuration` suffix whitelisted for `adapter.inbound.rest`
 * and its home here rather than in a new top-level `config` package (which the architecture tests
 * would reject).
 */
@RestControllerAdvice
class GlobalExceptionConfiguration {

    private val log = KotlinLogging.logger {}

    /** Invalid input — a bad UUID, a blank value object, an unknown status filter. */
    @ExceptionHandler(IllegalArgumentException::class)
    fun handleIllegalArgument(exception: IllegalArgumentException): ProblemDetail {
        log.debug { "Bad request: ${exception.message}" }
        return problem(HttpStatus.BAD_REQUEST, "Invalid request", exception.message)
    }

    /** Unknown aggregate — the services signal a missing application with [IllegalStateException]. */
    @ExceptionHandler(IllegalStateException::class)
    fun handleIllegalState(exception: IllegalStateException): ProblemDetail {
        log.debug { "Unprocessable: ${exception.message}" }
        return problem(HttpStatus.NOT_FOUND, "Resource not found", exception.message)
    }

    /** Action not available — the process token is no longer at the expected wait state. */
    @ExceptionHandler(MismatchingMessageCorrelationException::class)
    fun handleMismatchingCorrelation(exception: MismatchingMessageCorrelationException): ProblemDetail {
        log.debug { "Mismatching correlation: ${exception.message}" }
        return problem(HttpStatus.CONFLICT, "Action not available in the current state", exception.message)
    }

    private fun problem(status: HttpStatus, title: String, detail: String?): ProblemDetail =
        ProblemDetail.forStatus(status).apply {
            this.title = title
            this.detail = detail
            this.type = URI.create("https://miravelo.example/problems/${status.value()}")
        }
}
